tracer/src/Datadog.Trace/HttpOverStreams/ChunkedEncodingReadStream.cs

```diff
@@ -0,0 +1,323 @@
+                        // and try again
+                        await FillAsync().ConfigureAwait(false);
+                    }
+
+                    // We have a chunk header, so we can parse it
+                    _bytesRemainingInChunk = ParseChunkHexString(_streamBuffer, currentLine.Offset, currentLine.Length);
+
+                    // Proceed to handle the chunk.  If there's data in it, go read it.
+                    // Otherwise, finish handling the response.
+                    if (_bytesRemainingInChunk > 0)
+                    {
+                        _state = ParsingState.ExpectChunkData;
+                        break;
+                    }
+
+                    // We don't support trailers currently, so we expect another CRLF (chunk terminator)
+                    while (!TryReadNextChunkLine(out currentLine))
+                    {
+                        // We didn't have enough data to read the chunk terminator, so we need to refill the buffer
+                        // and try again
+                        await FillAsync().ConfigureAwait(false);
+                    }
+
+                    if (currentLine.Length != 0)
+                    {
+                        throw new Exception("Invalid response chunk terminator: expected 0 length terminator, received " + Encoding.ASCII.GetString(_streamBuffer, currentLine.Offset, currentLine.Length));
+                    }
+
+                    // all done!
+                    _state = ParsingState.Done;
+                    return 0;
+
+                case ParsingState.ExpectChunkData:
+                    // Read and return the chunk bytes
+                    // This reads the data from the stream directly into the provided buffer, instead
+                    // of into the _streamBuffer
+
+                    // As an optimization, we skip going through the connection's read buffer if both
+                    // the remaining chunk data and the buffer are both at least as large
+                    // as the connection buffer.  That avoids an unnecessary copy while still reading
+                    // the maximum amount we'd otherwise read at a time.
+                    if (_currentPosition.Length == 0
+                     && count >= ReadBufferSize
+                     && _bytesRemainingInChunk >= ReadBufferSize)
+                    {
+                        // We don't have enough for the whole chunk, so just read what we can, directly into the output buffer
+                        var bytesToRead = (int)Math.Min((ulong)count, _bytesRemainingInChunk);
+                        var bytesReadFromStream = await _innerStream.ReadAsync(buffer, offset, bytesToRead, cancellationToken).ConfigureAwait(false);
+
+                        // we might not have read the whole expected values, so just return what we have
+                        _bytesRemainingInChunk -= (ulong)bytesReadFromStream;
+                        if (_bytesRemainingInChunk <= 0)
```

Comment:
```suggestion
                        if (_bytesRemainingInChunk == 0)
```

If there's any suspicion that `_bytesRemainingInChunk ` can be negative here (by reading the code I don't see how) then we should throw